### PR TITLE
Allow scope admins to set property field permission levels

### PR DESCRIPTION
**Pull request by mgdelacroix**
#### Summary

Property field permission levels (`permission_field`, `permission_values`, `permission_options`)
could only be set at creation by system admins. Every other caller — including channel and team
admins — had all three silently overwritten with the object type's default, receiving **201 with
no indication the configuration was discarded**.

This blocked the Post Attributes shape where a channel admin defines a field for their channel
and controls who fills it in.

**What changed**

1. **Scope admins may configure permission levels.** The gate moves from `manage_system` to the
   admin of the field's own scope: `manage_system` for system-scoped fields, `manage_team` for
   team-scoped, `manage_channel_roles` for channel-scoped. It delegates to the same resolution
   `PermissionLevelAdmin` already uses, so "who may configure levels" and "who `admin` grants"
   cannot drift apart. Callers who don't administer the scope keep today's behaviour.

2. **DMs and group channels resolve `admin` to membership.** They have no channel-admin tier and
   cannot acquire one — member rows are only ever `SchemeUser`/`SchemeGuest`, and with no team
   there is no team scheme to inherit from. Without this, a participant could create a field and
   then be unable to edit, delete, or configure it. Applied in the shared resolver rather than the
   create gate, so the field definition and its values agree. **Guests are excluded**, matching the
   four channel-bookmark write handlers; they keep everything the `member` level grants, just as
   they can still list bookmarks.

3. **A field must be editable by its creator.** Creating a field with a `permission_field` the
   caller cannot satisfy is refused with 403, rather than producing a field only a system admin can
   change. Expressed as "the edit predicate must hold" and names no permission level, so it needs
   no change when new levels are added. Template-linked fields are exempt, since the app layer
   replaces their levels with the source template's.

**QA steps**

1. As a **channel admin**, `POST /api/v4/properties/groups/:group/fields/post` with
   `target_type: channel`, your channel's `target_id`, and `permission_field: admin` → **201**,
   levels honoured. On `master` this returns 201 with all three rewritten to `member`.
2. As a **plain channel member**, same request → **201**, levels replaced with `member` (unchanged).
3. As a **channel admin**, pin `permission_field: sysadmin` → **403**
   `api.property_field.create.creator_cannot_edit.app_error`.
4. As a **DM participant**, create a field scoped to that DM with `permission_field: admin` →
   **201**; then patch it, manage its options, set values, and delete it → all succeed. The other
   participant can too; a non-participant gets 403.
5. As a **guest** in a group channel, repeat (4) → refused the `admin` tier, but still able to read
   and to use `member`-level fields.
6. Confirm a plain member of a **normal** channel is still refused all of the above.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-70104

#### Release Note

```release-note
NONE
```

## Change Impact: 🔴 High

**Regression Risk:** The changes modify shared authorization logic and property-field access across API and session flows. Extensive automated tests reduce risk, but permission edge cases can affect access control.

**QA Recommendation:** Perform focused manual QA for system, team, channel, DM, group-channel, guest, and non-participant permissions.

*Generated by CodeRabbitAI*

### CODE REVIEW

File: server/channels/app/authorization.go [hunk 1]
```diff
@@ -602,10 +602,63 @@ func (a *App) HasPermissionToManagePropertyFieldOptions(rctx request.CTX, userID
...
+// rather than restating the cascade, so "who may configure a field's permission
+// levels" and "who the admin level grants" cannot drift apart.
+//
+// Returns false for a nil or protected field.
+func (a *App) SessionHasPermissionToAdministerPropertyFieldScope(rctx request.CTX, session model.Session, field *model.PropertyField) bool {
+	if field == nil || field.Protected {
+		return false
+	}
+	return a.hasPropertyFieldPermissionLevel(rctx, session.UserId, field, model.PermissionLevelAdmin)
+}
```
[the remark this answers is not in the document]
Comment (mgdelacroix) [reply server/channels/app/authorization.go:621]: There are no local API endpoints registered for the properties endpoints, so they're not exercised by the unrestricted sessions

**Review by davidkrauser — APPROVED** [review]
Clear and easy to follow change. Nice work 👏

**Review by JulienTant — APPROVED** [review]

File: server/channels/app/authorization.go [hunk 2]
```diff
@@ -602,10 +602,63 @@ func (a *App) HasPermissionToManagePropertyFieldOptions(rctx request.CTX, userID
...
+// property purposes: manage_channel_roles normally, or non-guest membership on
+// a DM/GM. A DM/GM has no channel-admin tier and cannot acquire one — its
+// member rows are only ever SchemeUser/SchemeGuest, and with no TeamId there is
+// no team scheme to inherit from — so its participants are its administrators,
+// matching how header edits and channel bookmarks are gated there.
+//
+// Guests are excluded. Administering a field is a write capability — it gates
+// editing and deleting the field and managing its options. Guests keep read
+// access through the member level.
+func (a *App) hasChannelPropertyAdmin(rctx request.CTX, userID, channelID string) bool {
```
Comment (edgarbellot) [inline server/channels/app/authorization.go:633]: `hasChannelPropertyAdmin` treats any non-guest DM/GM participant as "admin" for property fields. This is used in the shared resolver, so it also applies to fields that aren't scoped to that DM/GM - e.g. a system-level field with `permission_values: admin` (the shape Classification Markings uses).

Verified: a plain member with no channel role can now write an `admin`-level value on a DM/GM (200), where the same request returns 403 on master and in a public channel. Reverting just this PR's diff restores the 403.

Impact: any DM/GM participant can now set values on admin-only fields defined by a system/channel admin elsewhere - e.g. change a conversation's classification marking.

If this is not intentional, I would suggest to scope the DM/GM fallback to fields whose own target is that conversation (`field.TargetType == "channel"` and `field.TargetID` == the DM/GM), not any field whose value is recorded against that channel.

File: server/channels/api4/properties.go [hunk 3]
```diff
@@ -137,11 +137,16 @@ func createPropertyField(c *Context, w http.ResponseWriter, r *http.Request) {
...
-	isAdmin := c.App.SessionHasPermissionTo(*c.AppContext.Session(), model.PermissionManageSystem)
+	// Set before the permission levels are resolved.
+	field.CreatedBy = c.AppContext.Session().UserId
+	field.UpdatedBy = c.AppContext.Session().UserId
+
+	// Default permission levels: whoever administers the field's scope may
+	// configure them, everyone else has all three pinned to the object type's
+	// default.
+	canPin := c.App.SessionHasPermissionToAdministerPropertyFieldScope(c.AppContext, *c.AppContext.Session(), field)
 	defaultLevel := app.DefaultPropertyFieldPermissionLevel(field)
```
Comment (edgarbellot) [inline server/channels/api4/properties.go:148]: Not caused by this PR - but flagging since it's in code this PR touches.

A field with `object_type: post`/`channel` + `target_type: system` and omitted permission levels defaults to `member`. `member` on `target_type: system` means "any authenticated user." Verified: a plain member can PATCH (rename) and DELETE such a field, both 200.

Is this intentional? If not `DefaultPropertyFieldPermissionLevel` could default to `sysadmin` when `TargetType == "system"`, same reasoning `CanonicalizeSystemObjectField` already applies for `object_type: system`.

**Review by edgarbellot — CHANGES_REQUESTED** [review]
@mgdelacroix great work! Just two comments/questions :)

File: server/channels/app/authorization.go (hunk 2 above)
Comment (mgdelacroix) [reply server/channels/app/authorization.go:633]: This one is intentional. The rationale here is that participants act as admins within DMs/GMs, and they should be able to label their own conversations and manage them. If a system field has a `permission_values: admin` for posts on all channels within the platform, DM/GM users should be able to set these values.

This is only valid for values tho, a system field with `permission_field: admin` should not be editable by a DM/GM participant, and these changes should not allow it either.

Comment (mgdelacroix) [reply server/channels/app/authorization.go:633]: I've added a test on f6c7375a39 to pin the fact that the admin consideration should only affect property values, and never grant admin permissions to the field itself or its options

File: server/channels/api4/properties.go (hunk 3 above)
Comment (mgdelacroix) [reply server/channels/api4/properties.go:214]: Agree, good catch! Addressed in this PR in https://github.com/mattermost/mattermost/pull/38119/commits/65e408ff099cb00cc33dd11b4ec5bef46cfd4771, PTAL

File: server/channels/app/authorization.go (hunk 2 above)
Comment (edgarbellot) [reply server/channels/app/authorization.go]: Thanks for clarifying!

Noting CI is currently red: `TestPatchPropertyValuesChannelTargetAccess` and `TestLinkedProperties` (both on master, untouched here) assert the opposite of "DM/GM participants act as admins for values". I understand this will get synced before merge. Nothing else from security-wise, approving the PR :)

**Comment by mgdelacroix** [comment]
@davidkrauser @edgarbellot rerequesting review because I've had to add some permission changes to this PR. The recent inclusion of Channel Attributes in `master` encoded a rule in the generic PSAv2 permissions that forbid DM/GM participant to write attributes as per the future channel markings implementation, that expects them to be derived. This rule is specific from Channel Markings and drifts away from the generic, more permissive rule that Post Attributes relied on.

I've pushed f6ea7b9, which moves the more restrictive rule from the generic permission checks to the AccessControl hook, keeping the Channel Attributes functionality intact while encoding the generic rule of "DM/GM participants earn both member and admin tiers" in the generic codepaths of PSAv2. Please take a look at the commit (the rest of the PR is intact) and its body, which explains the rationale and mechanisms of the change.

Adding @pvev as well to ensure this doesn't change how Channel Attributes is expected to work.

File: server/channels/api4/properties.go (hunk 3 above)
Comment (edgarbellot) [reply server/channels/api4/properties.go:214]: Thanks for the hook split too, that resolves the DM/GM point.

On this fix: I think the sysadmin default is too broad, it also covers `permission_values`. `TestLinkedProperties` (on master, untouched here) expects linked fields to default `PermissionValues` to member and now fails. The hole I reported was only in the definition slots: `permission_field` and `permission_options` resolve via `hasPropertyFieldScopeAccess`, where a system target means any authenticated user. Values are always gated by the value's own target. Defaulting only field and options to sysadmin keeps both green.

Everything else LGTM :)

**Review by edgarbellot — APPROVED** [review]

File: server/channels/api4/properties.go (hunk 3 above)
Comment (mgdelacroix) [reply server/channels/api4/properties.go:214]: Fixed in https://github.com/mattermost/mattermost/pull/38119/changes/8bc36ccda53da9c6063e75f7d0d47bfd6a27ecae, PTAL :)

Comment (edgarbellot) [reply server/channels/api4/properties.go:214]: All good 👍 thanks for addressing it!

**Review by pvev — APPROVED** [review]
🎉

**Review by davidkrauser — APPROVED** [review]
